### DISCUSSION TITLE
[ARCHIVED] Processes vs Flows vs ...

### DISCUSSION THREAD

**Discussion by azinit** [feature-sliced/documentation#20] [Archive]
Как называть процессы? 

Слишком громоздко

> Предыстория

![image](https://user-images.githubusercontent.com/42924400/109698031-fc6dc180-7b9f-11eb-858e-4c064dfffcb6.png)

![image](https://user-images.githubusercontent.com/42924400/109697994-f1b32c80-7b9f-11eb-9533-6c4bbce6aa42.png)

![image](https://user-images.githubusercontent.com/42924400/109698061-055e9300-7ba0-11eb-8765-6c8c4d7ce2fb.png)

![image](https://user-images.githubusercontent.com/42924400/109698080-0a234700-7ba0-11eb-91a6-13f43bf4b274.png)

**Comment by akaia-shadowfox** [comment]
Я за flow / flows, но подумаю какие ещё варианты могут быть

**Reply by azinit** [reply]
Я в целом тоже приветствую более короткие и понятные названия

Но тут чет пока как ни крути - "процессы" вродь как лучше остальных (на данный момент)

**Reply by azinit** [reply]
Описывают ситуацию хорошо и `use-cases` / `cases` (у меня обычно под тест-кейсом, юз-кейсов - подразумевается перечень каких-то шагов, что в принципе процесс собой и представляет)

**Reply by azinit** [reply]
В том числе @sergeysova тоже хорошо сказал, что под контроллерами уже много другого сейчас подразумевается

Да, эта абстракция может оркестрировать какой-то логикой

Но как минимум - когда разраб увидет в структуре такую папку - он вообще не поймет сходу, что эти контроллеры "контролируют"

**А процессы в этом плане прям оч четенько заходят**

**Reply by akaia-shadowfox** [reply]
Вообще понятие "flow" или "workflow" уже широко используется, и обозначают примерно то, что описал Сергей, иногда эти процессы происходят даже не над страницами приложения, а над несколькими сервисами или приложениями.

https://workflow.is, ifttt, Microsoft Flow

**Comment by sergeysova** [comment]
Я за процессы. Похоже на процессы в ос и бизнес-процессы

**Reply by azinit** [reply]
Но и тут соглашусь, что пока вродь процессы достаточно нормально описывают вещи

У меня пока больше вопросов к `entities`))
> TL;DR: И по назначению (зачем там вдруг стейт),  и по наименованию
https://github.com/feature-sliced/wiki/discussions/18#discussioncomment-422625

**Reply by akaia-shadowfox** [reply]
Аналогия с процессами в ОС может только помешать, потому что может быть воспринята как threads / workers, как что-то низкоуровневое

**Reply by azinit** [reply]
> Аналогия с процессами в ОС может только помешать, потому что может быть воспринята как threads / workers, как что-то низкоуровневое

Ну как по мне - мы поэтому их потоками и не называем)

Т.к. потоки уже более что-то атомарное и низкоуровневое

А процессы (если мне память не изменяет), как-то как раз могут оркестрировать

Понятно, что сама ОС в этом плане гораздо большим контролем располагает

Но если я правильно помню, есть процесс - и он всяческими усилиями пытается решить свою задачу/проблему (забрать достаточно ресурсов вычислительных)

Поэтому хз даж

**Reply by azinit** [reply]
Можно, конечно подобрать и аналоги
- pipelines
- workflows (привет @karina-drummer )


Но сложно пока чего-то весомого набросить в сторону процессов

**Reply by akaia-shadowfox** [reply]
- bus
- channels
- drivers
- listeners
- pubsub

**Reply by akaia-shadowfox** [reply]
@sergeysova @martis-git выскажитесь что думаете над этими вариантами. Может какой-то зацепит.

**Reply by azinit** [reply]
А можешь пож напомнить, поч `processes` не заходят тебе? 

Аналогии из ОС?

**Reply by azinit** [reply]
<details>
<summary>Ну давай по каждому</summary>


> bus

Ассоциируется с работой с глобальными событиями

> channels

Ассоциируется с соединениями, и тоже что-то вроде глобальных событий

> drivers

Ну тут вообще web-driver'а вспоминаются (и не только для веб)

> listeners

Observer & Subscription - оч общо тож

> pubsub

Ребус какой-то)

Но тож `subscription` почему-то вырисовывается

</details>

---

А так - увы ни одно из них увы с процессом каким-то не ассоциируется(

**Reply by azinit** [reply]
@karina-drummer

**Reply by akaia-shadowfox** [reply]
> А можешь пож напомнить, поч `processes` не заходят тебе?
> 
> Аналогии из ОС?

- Труднопроизносимо
- Концепция "процессов" довольно размытая
- Аналогия с ОС может ввести в заблуждение. К тому же, если уж и использовать эту аналогию, то больше подходят сервисы / службы / демоны
- Есть более лаконичные либо более конкретизированные варианты

**Reply by azinit** [reply]
<details>
<summary>@karina-drummer  Ну хизе...))</summary>

![image](https://user-images.githubusercontent.com/42924400/111131576-71220200-8589-11eb-994a-42eb73884098.png)

![image](https://user-images.githubusercontent.com/42924400/111131529-649da980-8589-11eb-9e29-d984551219b1.png)

![image](https://user-images.githubusercontent.com/42924400/111131693-8eef6700-8589-11eb-9a87-c93c4060b8de.png)

![image](https://user-images.githubusercontent.com/42924400/111131755-9f074680-8589-11eb-907d-afbde07a91b0.png)

</details>

**Reply by azinit** [reply]
**А вообще, если по пунктам пройтись нормально**

> Труднопроизносимо

> Есть более лаконичные либо более конкретизированные варианты

Ну вот я тоже задавался этим вопросом, но почему-то **ничего другого более подходящего и не более сложного - не приходило в голову**

---

> Аналогия с ОС может ввести в заблуждение. 

А тут думаю, если без шуток, надо с двух точек зрения посмотреть:
1. "Я не знаю как работает ОС / знаю оч поверхностно" - человек в таком случае и **не будет проводить параллели с тем, с чем он мало знаком**
2. "Я знаю как работает ОС как минимум на достаточном уровне" - человек в таком случае поймет, либо что *сравнивать не особо корректно*, либо *все равно увидит* [*некие сходства*](https://github.com/feature-sliced/wiki/discussions/20#discussioncomment-422753)

---

> Концепция "процессов" довольно размытая

> К тому же, если уж и использовать эту аналогию, то больше подходят сервисы / службы / демоны

Тут думаю @sergeysova лучше развернет, но если я правильно понял - мы пришли к концепции **"`processes` ~ процессы, юзкейсы"**
*(Он и [вчера об этом говорил](https://github.com/feature-sliced/wiki/discussions/65#discussioncomment-480835))*

А потому `services`/`daemons` и прочие - вообще не подходят 
*(да, мб они как-то подходят "по реализации", но не "по назначению и смыслу")*

**Reply by azinit** [reply]
@feature-sliced/core , что скажете?
